src/dragonball/src/vmm.rs

```diff
@@ -0,0 +1,215 @@
+                        service
+                            .run_vmm_action(&mut v, &mut event_mgr)
+                            .unwrap_or_else(|_| {
+                                warn!("got spurious notification from api thread");
+                            });
+                    }
+                    if v.event_ctx.exit_evt_triggered {
+                        info!("Gracefully terminated VMM control loop");
+                        return v.stop(EXIT_CODE_OK as i32);
+                    }
+                }
+                Err(e) => {
+                    error!("Abruptly exited VMM control loop: {:?}", e);
+                    if let EpollError::EpollMgr(dbs_utils::epoll_manager::Error::Epoll(e)) = e {
+                        if e.errno() == libc::EAGAIN || e.errno() == libc::EINTR {
+                            continue 'poll;
+                        }
+                    }
+                    return vmm.lock().unwrap().stop(EXIT_CODE_GENERIC_ERROR as i32);
+                }
+            }
+        }
+    }
+
+    /// Waits for all vCPUs to exit and terminates the Dragonball process.
+    fn stop(&mut self, exit_code: i32) -> i32 {
+        info!("Vmm is stopping.");
+        if let Some(vm) = self.get_vm_mut() {
+            if vm.is_vm_initialized() {
+                if let Err(e) = vm.remove_devices() {
+                    warn!("failed to remove devices: {:?}", e);
+                }
+
+                if let Err(e) = vm.reset_console() {
+                    warn!("Cannot set canonical mode for the terminal. {:?}", e);
+                }
+
+                // Now, we use exit_code instead of invoking _exit to
+                // terminate process, so all of vcpu threads should be stopped
+                // prior to vmm event loop.
+                match vm.vcpu_manager() {
+                    Ok(mut mgr) => {
+                        if let Err(e) = mgr.exit_all_vcpus() {
+                            warn!("Failed to exit vcpu thread. {:?}", e);
+                        }
+                    }
+                    Err(e) => warn!("Failed to get vcpu manager {:?}", e),
+                }
+
+                // save exit state to VM, instead of exit process.
+                vm.vm_exit(exit_code);
+            }
+        }
+
+        exit_code
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+
+    pub fn create_vmm_instance() -> Vmm {
+        let info = Arc::new(RwLock::new(InstanceInfo::default()));
+        let event_fd = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let seccomp_filter: BpfProgram = Vec::new();
+        let epoll_manager = EpollManager::default();
+
+        Vmm::new_with_epoll_manager(
+            info,
+            event_fd,
+            epoll_manager,
+            seccomp_filter.clone(),
+            seccomp_filter,
+            None,
+        )
+        .unwrap()
+    }
+
+    #[test]
+    fn test_create_vmm_instance() {
```

Comment:
This is a **huge** PR but this is the only new test fwics...? Please could you consider writing more unit tests to help us have confidence in this new code? I'd also suggest breaking up a lot of the functions which are imho too big.

Reply:
In order to reduce the complexity and size of this PR, we have stripped the unit tests from this PR. We'll create a PR (which is the next Dragonball PR) for the unit tests.

Reply:
still need some time to rebase the code:( will do that tomorrow.

Reply:
#4604